### DISCUSSION TITLE
websocket protocol: Service and Action support

### DISCUSSION THREAD

**Discussion by beasteers** [foxglove/community#614] [Q&A]
I was looking through `ws-protocol` and didn't see any mention of ros services in the docs, issues, or code (though I did see some in foxglove bridge). I was wondering if this is something that is on the roadmap for the websocket protocol? Service and action emulation are both something that rosbridge provides so I was curious where the discussion was on this.

Thanks!

**Comment by aditya2592** [comment]
There is a ROS2 [REP](https://discourse.ros.org/t/rep-2012-service-introspection/26079) effort for service introspection. I would imagine that needs to be completed for service and action support to be added to foxglove studio or MCAP

**Reply by beasteers** [reply]
Hmm interesting, that looks useful. 

So would it be fundamentally different then this then?

[roslibpy services](https://roslibpy.readthedocs.io/en/latest/examples.html#using-services)

[roslibpy actions](https://roslibpy.readthedocs.io/en/latest/examples.html#actions)

I'm curious how close foxglove bridge is to being a (more efficient) replacement for rosbridge / roslibpy (or if I'm thinking about it wrong)

**Comment by jtbandes** [comment]
We are currently working on adding support for services to the protocol and foxglove_bridge, which should be available within a few weeks. Foxglove Studio does not currently support actions with rosbridge, so we have not prioritized support for actions in foxglove_bridge, but may work on it depending on the demand.

**Reply by beasteers** [reply]
Very cool that's great to hear! Is there a PR to follow?

My two cents is that supporting actions with foxglove_bridge would be useful on it's own just so that it can serve as a complete drop-in for rosbridge.

But also supporting actions in Foxglove Studio could have some benefit - e.g. a useful general purpose node would be a record-replay node that records data and saves it in an s3 bucket or to local disk - then in the foxglove UI, you could have a button that triggers the record Action and then outputs the progress/recording statistics. A UI element that is a button with some raw output could even be a general way to visualize actions and goals.

**Reply by jtbandes** [reply]
- https://github.com/foxglove/ws-protocol/pull/344
- https://github.com/foxglove/ros-foxglove-bridge/issues/10

**Reply by jhurliman** [reply]
Support for calling services through the foxglove bridge has just landed (https://github.com/foxglove/studio/pull/5266). It requires the latest https://github.com/foxglove/ros-foxglove-bridge on the server side so you will need to build from source to start using services today.
